### ISSUE TITLE
Make JAXB API dependency optional

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13047]
The move towards Jakarta EE (see https://github.com/jOOQ/jOOQ/issues/9641) has been causing a few issues in user projects, including projects that do not necessarily need jOOQ to have Jakarta EE dependencies.

While the validation stuff can be opted out of (it's a code generation only feature), we have two Jakarta EE runtime dependencies:

- JPA (already optional)
- JAXB (mandatory)

The JAXB dependency is mandatory even if the implementation is optional, as we rely on the JAXB API to unmarshal XML documents even using `MiniJAXB`, our own simple JAXB implementation that implements only the parts *we* need.

If we could duplicate all the annotations to our own internal packages, then `MiniJAXB` could read those, instead of the official ones, and we could make the entire JAXB API dependency optional and provided. This would not break compatibility, but greatly simplify jOOQ usage in legacy projects that have complex dependency trees.

**Comment by lukaseder** [comment]
The `XmlAdapter` isn't an annotation, but an abstract class, which we'd have to avoid in implementations, so I guess that it's not that simple...

**Comment by gpor0** [comment]
Or maybe provide new kind of Jooq configuration using properties file instead of xml which will result in more "lighweight" library?

**Comment by lukaseder** [comment]
> Or maybe provide new kind of Jooq configuration using properties file instead of xml which will result in more "lighweight" library?

XML itself isn't a problem. The fact that jOOQ has been using JAXB (carefully) made for a very seamless Maven integration, and even the [`gradle-jooq-plugin`](https://github.com/etiennestuder/gradle-jooq-plugin) profited from using the JAXB annotation based approach, which ultimately is a way of representing XML.

There's no convenient or "lightweight" way to model nested data structures as property files. How would you configure e.g.

- `<forcedTypes/>`: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/
- `<syntheticObjects/>`: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-synthetic-objects/

In fact, there's even nothing bad about JAXB itself. It is very well designed and useful to jOOQ. But the dependency has moved over the past years, causing trouble in client projects, especially when new users try out jOOQ, so the idea here is to enable jOOQ to work *without* the dependency. It already works without the *implementation* dependency, now this issue would also make the *API* dependency optional.

The saying is, I believe: ["Don't throw the baby out with the bathwater"](https://en.wikipedia.org/wiki/Don%27t_throw_the_baby_out_with_the_bathwater)

**Comment by gpor0** [comment]
I believe it is possible to model it anyway but I agree on that is better to use XML. Is it possible to at least move JAXB dependency from Jooq to Jooq-generator?

**Comment by lukaseder** [comment]
> I believe it is possible to model it anyway

Sure it is *possible*, but:

- It would be a lot of work
- It would be incompatible
- It would be unintuitive
- It would have to be documented very well
- People would hate migrating to it
- It would feel like busywork
- Why would we do it in the first place?

> Is it possible to at least move JAXB dependency from Jooq to Jooq-generator?

The runtime uses JAXB for `Settings` and `InformationSchema`: https://blog.jooq.org/jooq-3-16-and-java-ee-vs-jakarta-ee. It isn't there by accident...

We cannot *remove* JAXB without breaking things. But perhaps we can make it *optional*. This issue is to explore making the JAXB API dependency optional. It would be easy if we only used the annotations, but unfortunately, we're also using the `XmlAdapter` abstract class.

**Comment by lukaseder** [comment]
Also, let's not overstate the importance of this issue here. Yes, things are currently getting painful in the Java ecosystem because of the Java EE to Jakarta EE rename. But:

- They're getting painful for *everyone* and 2022 is the year folks migrate (because of Spring Boot, mainly, I woudl say)
- After the migration, stuff around JAXB will stabilise again

So, perhaps, we can just sit this one out? If making the API dependency optional isn't easy, then jOOQ won't do it.

**Comment by filiphr** [comment]
I found this issue while reading the blog the https://blog.jooq.org/jooq-3-16-and-java-ee-vs-jakarta-ee/ blog post. I am not a user of jOOQ and I do not know it is used in jOOQ, so please take my comment as best intention.

An alternative to use JAXB for reading / writing XMLs can be the `javax.xml.stream.XMLStreamReader` and `javax.xml.stream.XMLStreamWriter` which are part of the JDK. This API is of course low level and it is more cumbersome to use since you need to manually traverse it to read from it and manually write out the output. So if JAXB is not part of the jOOQ public API and it is only used internally then it could easily be replaced using the `XMLStreamReader` / `XMLStreamWriter` approach. We are using this at Flowable to read / write XML files. If you are interested to see how it looks like you can have a look at [`BpmnXMLConverter`](https://github.com/flowable/flowable-engine/blob/main/modules/flowable-bpmn-converter/src/main/java/org/flowable/bpmn/converter/BpmnXMLConverter.java).

Perhaps you have already explored this path, in any case I wanted to share a possible option.

**Comment by lukaseder** [comment]
Thanks for your comment, @filiphr.

The implementation is not a problem. It is already optional. You can find our own "mini JAXB implementation" here:
https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/util/jaxb/tools/MiniJAXB.java

It just uses the DOM API. `XMLStreamReader` would have been an option, too, of course. But nothing needs to be changed in this area.

This issue here is about making the JAXB *API* optional.

**Comment by lukaseder** [comment]
An internal dependency on `DatatypeConverter` has been removed: https://github.com/jOOQ/jOOQ/issues/13393

**Comment by wcurrie** [comment]
Feedback from one (unpaid) user: we're current blocked from upgrade past 3.15 because other dependencies in our tree are still pegged to an older jaxb API version. We don't read `Settings` from XML. We just say `new Settings()`. We tried excluding the dependency with:
```xml
<exclusions>
  <exclusion>
    <groupId>jakarta.xml.bind</groupId>
    <artifactId>jakarta.xml.bind-api</artifactId>
  </exclusion>
</exclusions>
```

This seems to work but javac emits a warning about not being able to find the annotations referenced from `Settings`:
```
warning: unknown enum constant XmlAccessType.FIELD
```
We currently convert warnings to errors. It doesn't seem possible to suppress just that warning so all warnings would need to be disabled. Which is undesirable for us.

I'm assuming there's no such thing as jackson [mixins](https://github.com/FasterXML/jackson-docs/wiki/JacksonMixInAnnotations#mix-ins-to-the-rescue) for jaxb implementations that might provide a road to separating the `@XmlElement` into some separate+optional `jooq-jaxb` artefact.

I presume if it were easy to remove the dependency it'd be done already. Thanks.

**Comment by lukaseder** [comment]
> Feedback from one (unpaid) user: we're current blocked from upgrade past 3.15 because other dependencies in our tree are still pegged to an older jaxb API version

I'm really curious, apart from uncertainty and the obvious feeling that things shouldn't be this way: In principle, having both javax and jakarta on the classpath shouldn't be an issue. The namespaces are mutually exclusive, and in theory, the two JAXB versions are incompatible and shouldn't interfere. What's the problem with having both on the classpath?

> I'm assuming there's no such thing as jackson [mixins](https://github.com/FasterXML/jackson-docs/wiki/JacksonMixInAnnotations#mix-ins-to-the-rescue) for jaxb implementations that might provide a road to separating the `@XmlElement` into some separate+optional `jooq-jaxb` artefact.

That looks complicated. Not something that I wish to spend time on, given how unimportant this dependency (and the other jakarta dependencies) is for jOOQ

> I presume if it were easy to remove the dependency it'd be done already. Thanks.

We can't remove it as third parties rely on the annotations being present on:

- `Settings`
- `InformationSchema`
- `Configuration` (from the code generator)

People use `JAXB.marshal()` and `JAXB.unmarshal()` with this API, and when jOOQ removed the implementation dependency, we re-implemented those marshal/unmarshal methods partially in our own internal `MiniJAXB`, which is used occasionally behind the scenes.

I mean, JAXB *is* useful, and there hasn't been a very strong reason to get rid of it entirely.

This issue here is only about making the dependency optional in Maven (it already is optional (`requires static`) in the `module-info.java`)

**Comment by lukaseder** [comment]
Anyway, let me have another look at this optional dependency idea and what it means for users. It's hard to guess what happens without trying it in an actual build.

**Comment by lukaseder** [comment]
Yeah, I think that with the removal of the last non-annotation dependency as of https://github.com/jOOQ/jOOQ/issues/13047#issuecomment-1084540902, we can try making the dependency optional in the *runtime library*. I'm reluctant to make it optional in the code generator, as people probably use JAXB more often there, when configuring the code generator programmatically.

**Comment by lukaseder** [comment]
Alright, I don't really see any problems with this change, which consists of:

**pom.xml** (runtime library)

```patch
diff --git a/jOOQ/pom.xml b/jOOQ/pom.xml
index c1f8f8c..aaef52a 100644
--- a/jOOQ/pom.xml
+++ b/jOOQ/pom.xml
@@ -116,12 +116,17 @@
             <artifactId>jakarta.persistence-api</artifactId>
         </dependency>
 
-        <!-- From JDK 9 onwards, the JAXB dependency needs to be made explicit -->
-        <!-- The dependency can cause trouble in older JDKs, so it needs to be
-             excluded from pre-java-9 builds: https://github.com/jOOQ/jOOQ/issues/7649 -->
+        <!-- From JDK 9 onwards, the JAXB dependency needs to be made explicit
+        
+             [#7649] The dependency can cause trouble in older JDKs, so it needs to be
+             excluded from pre-java-9 builds: https://github.com/jOOQ/jOOQ/issues/7649
+             [#13047] The dependency is now optional. Users can opt in to using this
+             dependency if they need, by making it explicit on their end. 
+          -->
         <dependency>
             <groupId>jakarta.xml.bind</groupId>
             <artifactId>jakarta.xml.bind-api</artifactId>
+            <optional>true</optional>
         </dependency>
         
         <!-- [pro] -->
```

**pom.xml** (jOOQ-meta)

```patch
diff --git a/jOOQ-meta/pom.xml b/jOOQ-meta/pom.xml
index 9f51d8d..989e337 100644
--- a/jOOQ-meta/pom.xml
+++ b/jOOQ-meta/pom.xml
@@ -54,6 +54,13 @@
             <scope>provided</scope>
             <optional>true</optional>
         </dependency>
+        
+        <!-- [#13047] The dependency is optional in the runtime library, but
+                      there's no need to make it optional also for the code generation -->
+        <dependency>
+            <groupId>jakarta.xml.bind</groupId>
+            <artifactId>jakarta.xml.bind-api</artifactId>
+        </dependency>
 
         <!-- [test] -->
         <dependency>
```

I'm a bit conservative here with `jOOQ-meta`. All of `jOOQ-meta`, `jOOQ-codegen`, `jOOQ-codegen-maven`, as well as third parties like https://github.com/etiennestuder/gradle-jooq-plugin rely on these annotations. Making them optional is asking for trouble, and again, I think people are using JAXB with the code generator because the jOOQ user manual has documented this for ages:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-programmatic/

If this turns out to be a big problem, we can still address it, but I think most complaints stem from the runtime library having this dependency.

The dependency won't be provided, because we use the annotations internally, even when users don't provide the dependency explicitly, so it's not like an SPI or a JDBC driver. `MiniJAXB` doens't work when the annotations can't be loaded.

I'll push a commit soon, which will be integrated in the next 3.18.0-SNAPSHOT build available from https://www.jooq.org/download/versions or directly from github. Early feedback would be greatly appreciated!

**Comment by lukaseder** [comment]
> This seems to work but javac emits a warning about not being able to find the annotations referenced from `Settings`:
> 
> ```
> warning: unknown enum constant XmlAccessType.FIELD
> ```

I'm curious about this warning, which I couldn't reproduce. How can it be triggered?

**Comment by wcurrie** [comment]
Thanks for the super rapid response as always.

> What's the problem with having both on the classpath?

🤦 I'd missed that the artifact id changed from `javax.xml.bind:jaxb-api` to `jakarta.xml.bind:jakarta.xml.bind-api`. So I'd assumed this wasn't possible. Thanks! I'll try. I reckon you've unblocked our upgrade just by pointing that out.

<img width="1595" alt="image" src="https://user-images.githubusercontent.com/2050920/189334895-ad57b230-7c58-4f44-8930-a75656edb7b9.png">

> I'm curious about this warning, which I couldn't reproduce. How can it be triggered?

I'm off on vacation for a while sorry. I'll try to reproduce when I'm back.

**Comment by lukaseder** [comment]
A reproducer has been provided here: https://github.com/jOOQ/jOOQ/issues/14865. Will look again into this problem and report findings in that issue.

**Comment by wcurrie** [comment]
Some background to the problem we face: 

* we have dependencies (other than jooq) relying on `package javax.xml` used by jakarta.xml.bind:jakarta.xml.bind-api version 2.3.0.
* jakarta.xml.bind:jakarta.xml.bind-api version 3.0.0 flips to using `package jakarta.xml`. "this is fucking nightmare for everyone" in [1] seems on point.

So I think we either need to:
* upgrade everything in our dependency tree to jakarta.xml.bind:jakarta.xml.bind-api version 3.0.0 (simultaneously)
* find everything that today is pulling in jakarta.xml.bind:jakarta.xml.bind-api version 2.3.0 (which will flip to package jakarta.xml in 3.x, breaking random things) and make it pull in the older javax.xml.bind:jaxb-api (which uses package javax.xml)
* find a way to cope with https://github.com/jOOQ/jOOQ/issues/14865. which seems to be: ignore all javac warnings. And use a separate dependency tree/closure when running jooq codegen (which needs the `package jarkarta.xml` dependency)

Jackson forked [1] the components that depend on `package javax.xml` or `package jakarta.xml` 

We don't use gradle, otherwise we'd try something like https://blog.ltgt.net/javax-jakarta-mess-and-gradle-solution/

<img width="677" alt="image" src="https://github.com/jOOQ/jOOQ/assets/2050920/5aa02a5a-335c-42e1-be8e-d32ad9ac8b22">

We probably need to work out a way to achieve ☝️ in our build system.

I'd be stoked to learn I'm daft and missing a simpler solution.

[1] https://groups.google.com/g/jackson-dev/c/mlPkj7cWfd8

**Comment by lukaseder** [comment]
> * find a way to cope with [Compiler warning for XmlAccessType with jOOQ 3.18.0 #14865](https://github.com/jOOQ/jOOQ/issues/14865). which seems to be: ignore all javac warnings.

I think this particular flaw (I still have the tab open: https://bugs.openjdk.org/browse/JDK-8305250) just shows that `-Werror` is very broken. There are sophisticated static code analysers out there. `javac` is not one of them. Irrespective of whether this particular bug will be accepted (and fixed!) by the JDK teams, the fact that you can't opt out of individual, very specific warnings has always been a mystery to me.

I recommend using something else than `javac` warnings for code quality.

> * And use a separate dependency tree/closure when running jooq codegen (which needs the `package jarkarta.xml` dependency)

I'm assuming you now need this because of the third party plugin by https://github.com/etiennestuder/gradle-jooq-plugin ? His Gradle/Groovy integration does indeed read the JAXB annotations, I think. If you're using Gradle/Kotlin, then you won't have this requirement.

> I'd be stoked to learn I'm daft and missing a simpler solution.

I don't think there is. This "simple" namespace change affects the entire ecosystem, fundamentally, in parts because JAXB was once part of the JDK (since Java 6), not just a Java EE module. If that hadn't been the case, then jOOQ probably would not have had this dependency.
